finagle-netty4-http/src/main/scala/com/twitter/finagle/netty4/http/package.scala

```diff
@@ -146,7 +147,7 @@ package object http {
         // specific HttpServerExpectContinueHandler above.
         pipeline.addLast(
           "fixedLenAggregator",
-          new FixedLengthMessageAggregator(maxRequestSize, handleExpectContinue = false)
+          new FixedLengthMessageAggregator(maxRequestSize, dechunkThreshold, handleExpectContinue = false)
```

Comment:
I had no intention to implement client side, but I'm open doing this if you think we need that. I do not know, however, what is the behavior right now, and what it should be. Would need to investigate.

`minChunkSize` is a bit confusing for me.
In general, it is confusing, that finagle's chunk is different from http chunk. It took me some time to realize, that `Request.isChunked` has, strictly speaking, nothing to do with `Transfer-Encoding` header.
If you believe, that for experienced finagle users `minChunkSize` would clearly mean finagle's chunk, then let's do this name. Otherwise, I, as non-experienced finagle user, would prefer to see the parameter named smthng like `streamingThreshold`

Reply:
Let's go with `minChunkSize` for now, if we think of something better during the review we can always change it.

Reply:
I think we should change the client-side too as, otherwise, `Http.client.withStreaming(true, 128.kb)` wouldn't mean the `minChunkSize` is silently ignored.

Reply:
I implemented your suggestions on the server-side (had to force push — I noticed that my feature branch been branched from the master).

I need some time to figure out how exactly the client-side should be changed.

Reply:
@vkostyukov , just to clarify.

It seems that in your last message you assume that we already did changes to API but did not modify pipeline creation code.

While in fact I didn't change the API. Signature of the`Http.client.withStreaming` is the same as before this PR: `withStreaming(Boolean)`.  

> Http.client.withStreaming(true, 128.kb) would mean the minChunkSize is silently ignored.

Is this your only concern? Or do you indeed want to change client's behavior?

Reply:
Oh you're right. I miss-remembered that `withStreaming`  configuration API is shared between client and server. I think that's fine then. It's ok not to be symmetric there (looks like we never were).

Perhaps wait for more people to chime in w.r.t. client-side configuration. 👍 from me for just changing a server for now.

Reply:
Given a client `Http.client.withStreaming(true, minChunkSize = 128.kilobytes)`, what does it do when it receives a response from the server with `Content-Length: 512KB`?

Does it mean that given `loop`,

```
def loop: Future[Unit] =
  response.reader.flatMap {
    case Some(buf) => go(buf).before { loop }
    case None => Future.Done
  }
```

, `go(buf)` will never receive a `buf` where `buf.length > 128.kilobytes`?

Also, what happens if the server responds with `Transfer-Encoding: chunked`, no `Content-Length`, and sends chunks that are only 4.kilobytes, e.g.:

```
Transfer-Encoding: chunked

4096\r\n
[...4096 chars...]\r\n
```

Do we buffer these 4KB chunks into 128KB chunks for the client?

Reply:
@luciferous, I don't know the answer, but I'll reply just because I want to understand this too and I hope someone will correct me if I'm wrong.

First of all, I don't think there's `flatMap` method on `Reader`, you likely missed the call to `read` in you code.

Read requires an argument ­— number of bytes to read. So given the loop
```
    def loop: Future[Unit] = {
      req.reader.read(128.kilobytes).flatMap {
        case Some(buf: Buf) => go(buf).before(loop)
        case None =>  uture.Done
      }
    }
```

* response `Content-Length` is `<= 128.kilobytes`
`go(buf)` will receive `buf.length <= 128.kilobytes` as soon as _whole_ message is read. The loop will have exactly one iteration.
* response `Content-length` is `> 128.kilobytes`
`go(buf)` will receive buf of an _unknown (to me)_ size defined by netty pipeline and/or sending party. The loop will have multiple itrations.
* response `Transfer-Encoding: chunked` is set
`go(buf)` will receive buf of an _unknown (to me)_ size defined by netty pipeline and/or sending party. The loop will have multiple iterations.

IDK if my understanding is correct, but this is exactly my concern with the name `minChunkSize`. Although http chunk size (or any chunk size) per se, is not exposed via finagle API, parameter name suggests (at leat to me), that there's some buffering in the middle, and that the consumer of the `reader` will see the data only after `minChunkSize` requirement has been met (or on EOF). I'm not pushing for a name change, just trying to be on the same page with you folks.

Reply:
It seems to me what no matter how we name it, there is still some buffering happens (at least today) for messages with known `Content-Length`. We can surely document what `minChunkSize` applies to. I'd imagine, setting `minChunkSize = 0` would effectively mean "no buffering". I'm trying to understand your reasoning and I'm still not sure why you don't like the name. It does what it says it does (for messages with known length - is this part a concern?).

Reply:
> First of all, I don't think there's flatMap method on Reader, you likely missed the call to read in you code.
@edio Yes, that correct, sorry for any confusion.

> IDK if my understanding is correct, but this is exactly my concern with the name minChunkSize.
@edio Mine too.

> I'm trying to understand your reasoning and I'm still not sure why you don't like the name. It does what it says it does (for messages with known length - is this part a concern?).
@vkostyukov It's confusing because calling `Reader.read` can return chunks larger than `minChunkSize`.

Sorry, I know I changed my mind on the name 😅. What if we called it `bufferSize`? Then:
1. `bufferSize = 0`: no buffer, `Reader.read` emits exactly what comes out of the Netty pipeline
2. `bufferSize = 128.kilobytes`: buffer up to 128.kilobytes; `Reader.read` will ALWAYS return `buf` where `buf.length <= 128.kilobytes`
3. We also note that this parameter doesn't apply to the "write" path, i.e. when the user does a `writer.write(buf)`, `buf.length` is unrestricted here.

Reply:
@vkostyukov, buffering indeed may happen, but not in `FixedLengthMessageAggregator`.
Aggregator's buffering is all-or-nothing: we either buffer all message into a huge buffer up to `maxContentLength` or we by pass the aggregator, and emit chunks of data (and in this case `minChunkSize` does not have any effect).

So for example if we have a sufficiently large `minChunkSize` (say, 128 KiB) that is used as a `maxContentLength` for `FixedLengthMessageAggregator`, then the consumer of `Request.reader` wil actually never read a buffer of size 128*1024. On contrary, 128 KiB will likely be maximal theoretical value the consumer can read.

Reply:
Okay, I'm convinced. Sorry for the confusion, I think I mis-understood what `FixedLengthMessageAggregator` does. `bufferSize` seems reasonable but, still, could be interpreted as "buffer my chunks pls".

Perhaps something as verbose as `aggregateIfLessThan = 128.kb`.

Reply:
I don't have any better ideas. `aggregateIfLessThan` works for me.

Reply:
@luciferous 
small clarification
> `bufferSize = 128.kilobytes`: buffer up to 128.kilobytes; `Reader.read` will ALWAYS return `buf` where `buf.length <= 128.kilobytes`

if 
* bufferSizez is very small. For example `bufferSize == 100.bytes`
* request has `Transfer-Encoding: chunked` or request `Content-Length` is `> 100`
then `Request.reader.read(128.kilobytes.inBytes)` will likely read `buf` with `buf.length > 100` (the value will be determined by upstream handlers in netty pipeline)

if 
* bufferSizez is very small. For example `bufferSize == 100.bytes`
* request does NOT have `Transfer-Encoding: chunked` and request `Content-Length` is `<= 100`
then `Request.reader.read(128.kilobytes.inBytes)` is guaranteed to read `buf` with `buf.length <= 100`

So, as you see, there are no guarantees at all.

This discussion makes me think, that maybe we need couple of tests to demonstrate that behavior, so it is easier to reason about the handler behavior?
In the meantime I will make an attempt to improve the scaladoc on the aggregator.

Reply:
OK, I'm not sure I follow, but hopefully some test/docs will clarify. I'll think about it more too.